pinot-transport/src/main/java/com/linkedin/pinot/routing/TableConfigRoutingTableSelector.java

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.routing;
+
+import com.google.common.base.Splitter;
+import com.linkedin.pinot.common.config.RealtimeTableConfig;
+import com.linkedin.pinot.common.metadata.ZKMetadataProvider;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.commons.configuration.Configuration;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Routing table selector that uses the "routing.llc.percentage" (between 0-100) custom config in metadata to determine
+ * the appropriate ratio of queries to route to LLC.
+ */
+public class TableConfigRoutingTableSelector implements RoutingTableSelector, IZkDataListener {
```

Comment:
Most of the complexity is from reading a value from ZK and caching it, while trying to avoid having a stale ratio value for too long. A simpler implementation would forego the cache and read the config value from ZK on every query, which is not what we want.

How would you make it simpler?